src/AzureClient/AzureClient.cs

```diff
@@ -178,38 +183,45 @@ private async Task<ExecutionResult> SubmitOrExecuteJobAsync(IChannel channel, Az
             {
                 return await GetJobStatusAsync(channel, MostRecentJobId);
             }
 
+            // If the command was %azure.execute, wait for the job to complete and return the job output.
             channel.Stdout($"Waiting up to {submissionContext.ExecutionTimeout} seconds for Azure Quantum job to complete...");
 
-            using (var cts = new System.Threading.CancellationTokenSource(TimeSpan.FromSeconds(submissionContext.ExecutionTimeout)))
+            using (var executionTimeoutTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(submissionContext.ExecutionTimeout)))
+            using (var executionCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(executionTimeoutTokenSource.Token, cancellationToken))
             {
-                CloudJob? cloudJob = null;
-                do
+                try
                 {
-                    // TODO: Once jupyter-core supports interrupt requests (https://github.com/microsoft/jupyter-core/issues/55),
-                    //       handle Jupyter kernel interrupt here and break out of this loop 
-                    await Task.Delay(TimeSpan.FromSeconds(submissionContext.ExecutionPollingInterval));
-                    if (cts.IsCancellationRequested) break;
-                    cloudJob = await ActiveWorkspace.GetJobAsync(MostRecentJobId);
-                    channel.Stdout($"[{DateTime.Now.ToLongTimeString()}] Current job status: {cloudJob?.Status ?? "Unknown"}");
+                    CloudJob? cloudJob = null;
+                    while (cloudJob == null || cloudJob.InProgress)
+                    {
+                        executionCancellationTokenSource.Token.ThrowIfCancellationRequested();
+                        await Task.Delay(TimeSpan.FromSeconds(submissionContext.ExecutionPollingInterval), executionCancellationTokenSource.Token);
+                        cloudJob = await ActiveWorkspace.GetJobAsync(MostRecentJobId);
+                        channel.Stdout($"[{DateTime.Now.ToLongTimeString()}] Current job status: {cloudJob?.Status ?? "Unknown"}");
+                    }
+                }
+                catch (Exception e) when (e is TaskCanceledException || e is OperationCanceledException)
+                {
+                    Logger?.LogInformation($"Operation canceled while waiting for job execution to complete: {e.Message}");
                 }
-                while (cloudJob == null || cloudJob.InProgress);
             }
 
             return await GetJobResultAsync(channel, MostRecentJobId);
         }
 
         /// <inheritdoc/>
-        public async Task<ExecutionResult> SubmitJobAsync(IChannel channel, AzureSubmissionContext submissionContext) =>
-            await SubmitOrExecuteJobAsync(channel, submissionContext, execute: false);
+        public async Task<ExecutionResult> SubmitJobAsync(IChannel channel, AzureSubmissionContext submissionContext, CancellationToken cancellationToken) =>
```

Comment:
Since this is public, it may be a nice convienence to have `cancellationToken` default to `CancellationToken.None`.

Reply:
Yes, that's a good idea. It looks like I can't use `CancellationToken.None` as a default value because it's not a compile-time constant, but I'll make this take a `CancellationToken?`, default to `null`, and have the code use `cancellationToken ?? CancellationToken.None`.